pkgs/dart_tooling_mcp_server/lib/src/mixins/pub_dev_search.dart

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:dart_mcp/server.dart';
+import 'package:http/http.dart';
+import 'package:pool/pool.dart';
+
+import '../utils/json.dart';
+
+// Override this to stub responses for testing.
+Client Function() createClient = Client.new;
+
+/// Limit the number of concurrent requests.
+final _pool = Pool(10);
+
+/// The number of reults to return for a query.
+// If this should be set higher than 10 we need to implement paging of the
+// http://pub.dev/api/search endpoint.
+final _resultsLimit = 10;
+
+/// Mix this in to any MCPServer to add support for doing searches on pub.dev.
+base mixin PubDevSupport on ToolsSupport {
+  @override
+  FutureOr<InitializeResult> initialize(InitializeRequest request) {
+    registerTool(pubDevTool, _runPubDevSearch);
+    return super.initialize(request);
+  }
+
+  /// Implementation of the [pubDevTool].
+  Future<CallToolResult> _runPubDevSearch(CallToolRequest request) async {
+    final query = request.arguments?['query'] as String?;
+    if (query == null) {
+      return CallToolResult(
+        content: [
+          TextContent(text: 'Missing required argument `search-query`.'),
+        ],
+        isError: true,
+      );
+    }
+    final client = createClient();
+    final searchUrl = Uri(
+      scheme: 'https',
+      host: 'pub.dev',
+      path: 'api/search',
+      queryParameters: {'q': query},
+    );
+    final Object? result;
+    try {
+      result = jsonDecode(await client.read(searchUrl));
+
+      final packageNames =
+          dig<List>(result, ['packages'])
+              .take(_resultsLimit)
+              .map((p) => dig<String>(p, ['package']))
+              .toList();
+
+      if (packageNames.isEmpty) {
+        return CallToolResult(
+          content: [
+            TextContent(
+              text: 'No packages mached the query, consider simplifying it.',
+            ),
+          ],
+          isError: true,
+        );
+      }
+
+      Future<Object?> retrieve(String path) async {
+        return _pool.withResource(() async {
```

Comment:
[nit] We aren't using the outer `async` since there is no `await`.

```suggestion
      Future<Object?> retrieve(String path) {
        return _pool.withResource(() async {
```

(alternatively we would use `return await` if we did need the `async`)